### ISSUE TITLE
List of maps is behaving strange

### ISSUE THREAD

**Issue by zbikmarc** [hashicorp/terraform#14309]
We returning map as one of outputs. Then we are getting this data using `data "terraform_remote_state"`. This should create list of maps, containing IPs as keys and dns names as values. However it looks like it crates some strange data structure where IPs are omitted at all, and there is some strange nesting.

### Terraform Version
`Terraform v0.9.2`

### Terraform Configuration Files
```hcl
data "terraform_remote_state" "remote-envs" {
  count   = "${compute_number}"
  (backend configuration here)
}

output "output" {
  value = "${list(data.terraform_remote_state.remote-envs.*.map-output)}"
}
```

### Relevant outputs
Single map:
```hcl
map-output = {
  192.168.1.100 = dns-name1
  192.168.1.101 = dns-name2
}
```

List of maps:
```hcl
debug = [
    [
        map[192:map[168:map[1:map[100:dns-name1 101:dns-name2]]]]
    ]
]
```

### Expected Behavior
Terraform should return list of maps, easy parsable one.

### Actual Behavior
As showed above, some strange structure is returned. Look at number on map... those are parts of IP... (we just found it).

### Additional info.
I tried to use this:
```hcl
output "output" {
  value = "${element(data.terraform_remote_state.remote-envs.*.map-output, 0)}"
}
```
but it does not even produce output at all.

**Comment by apparentlymart** [comment]
Hi @zbikmarc! Sorry for the problem here.

Funnily enough I was looking at a similar problem with someone on the Gitter chat yesterday. My theory (not yet confirmed) is that the special dynamic attribute behavior of `terraform_remote_state` causes problems for Terraform's "splat syntax".

`terraform_remote_state` is a rather unique resource in Terraform in that the set of attributes it exposes is decided dynamically based on the set of outputs it finds in the state. Usually Terraform wants to check that all attributes accessed are valid before evaluating an expression, but for this resource that isn't possible.

I've not yet delved into the code to see the full details here, but I assume that the codepath that handles the "splat syntax" doesn't know how to deal with the dynamic attributes on this resource, and so the expression `data.terraform_remote_state.remote-envs.*.map-output` is returning an error. This then interacts with another issue where errors in outputs are silently swallowed; if this expression where anywhere _other_ than an output I think you would see an error saying that this resource doesn't have an attribute called "map-output".

Unfortunately I don't have a ready workaround for this problem at this time.

**Comment by rgardam** [comment]
I am seeing something which I think is related. 

i'm setting a map as an output in my main.tf file for one set of states and then am trying to consume this map from the remote state. I am getting a "does not have attribute" even though I can see the exact output in the root level outputs of that state. 

Could it be that the terraform_remote_states provider doesn't handle maps correctly at all?

**Comment by stefano-m** [comment]
@apparentlymart I can confirm this behaviour and have a simple reproducer.

I have been able to reproduce the issue with the below versions of Terraform:

* Terraform 0.10.8
* Terraform 0.11.1

You can find the code and JSON outputs here: https://gist.github.com/stefano-m/360a0e3f5cba81ceead2a5b610395739

Also note that in the output that uses the remote state (`one.tf`) the value of the `"something"` key is a list rather than a map.

I am very interested in a solution to this issue and I'd be happy to help (with some guidance).

**Comment by apparentlymart** [comment]
Thanks for that additional information, @stefano-m!

The Terraform team at HashiCorp is currently working on integrating an improved version of the configuration language that will lead to some significant changes to the relevant code here, so to avoid conflicts we're going to look deeper into this problem once we've made more progress with the configuration language work.

I expect that the work already ongoing will improve the situation automatically just due to the different implementation of the splat syntax in the new version (it's now a proper operator in the language, rather than a special case for resource attributes) but I expect there will be some other followup changes required, which will be easier to see/test once we've reached that point.

**Comment by stefano-m** [comment]
@apparentlymart thanks for the reply.

I hope that eventually the reproducer could be used to prove that the issue has been fixed.

It took me a whole day to work around the bug in a hacky way and I look forward to having the fix available.

Thank you for working on Terraform.

**Comment by apparentlymart** [comment]
Hi all! Sorry for the long silence.

Over in #10876 I verified a problem with the same root cause against the v0.12.0-alpha2 prerelease build. This fix is already in the master branch and ready to be included in the forthcoming v0.12.0 release, so I'm going to close this out now.

Thanks for reporting this, and sorry for how long it took to work through all the layers to get it fixed.
